bigtable/client/internal/endian.h

```diff
@@ -0,0 +1,153 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ENDIAN_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ENDIAN_H_
+
+#include "bigtable/client/internal/throw_delegate.h"
+#include <bitset>
+#include <limits>
+#include <vector>
+
+#if defined(_MSC_VER)
+#include <stdlib.h>
+#else
+#include <byteswap.h>
+#endif
+
+#include <string>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+namespace internal {
+
+#if defined(_MSC_VER)
+#define byteswap64 _byteswap_uint64
+#define byteswap32 _byteswap_ulong
+#else
+#define byteswap64 __builtin_bswap64
+#define byteswap32 __builtin_bswap32
+#endif
+
+constexpr char c[sizeof(int)] = {1};
+constexpr int i = {1};
+constexpr bool IsBigEndian() {
+  return i != c[0];  // ignore different type comparison
+}
+
+typedef std::int64_t bigendian64_t;
+typedef std::int32_t bigendian32_t;
+
+template <typename T>
+struct encoder {
+  static std::string encode(T const& value);
+  static T decode(std::string const& value);
+};
+
+template <>
+struct encoder<bigendian64_t> {
+  /**
+   * This function will convert the number into its equivalent
+   * bigEndian 64 bit string of bytes.
+   * If char size is not 1 byte (8 bits) then
+   * Raise the exception.
+   *
+   * @param value reference to a number need to be converted into
+   *        bytes of 64 bit BigEndian hex string.
+   */
+  static std::string encode(bigendian64_t const& value) {
+    static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                  "This code assumes char is an 8-bit number");
+
+    std::string big_endian_string("", 8);
+    if (IsBigEndian()) {
+      std::memcpy((void*)big_endian_string.data(), &value, sizeof(value));
+    } else {
+      bigendian64_t swapped_value = 0;
+      swapped_value = byteswap64(value);
+      std::memcpy((void*)big_endian_string.data(), &swapped_value,
+                  sizeof(swapped_value));
+    }
+    return big_endian_string;
+  }
+
+  /**
+   * This function will convert bigEndian 64 bit string of bytes into
+   * into its equivalent numeric value.
+   * If value passed is not 8 bytes strings then this function will
+   * Raise the exception std::range_error.
+   *
+   * @param value reference to a number need to be converted into
+   *        bytes of 64 bit BigEndian hex string.
+   */
+  static bigendian64_t decode(std::string const& value) {
+    // Check if value is BigEndian 64-bit integer
+    if (value.size() != 8) {
+      internal::RaiseRangeError("Value is not convertible to uint64");
+    }
+    bigendian64_t big_endian_value = 0;
+    std::memcpy(&big_endian_value, value.c_str(), 8);
+    if (!IsBigEndian()) {
```

Comment:
I think we can trust compilers to optimize this code, I just tested:

https://godbolt.org/g/fQ3ALJ